cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+Additionally SDP also needs to translate IP address to endpoint ID and identity in order to enforce policy by reusing the logic from agent's DNS proxy. Our proposal involves retrieving the endpoint and identity data directly from the `cilium_lxc` and `cilium_ipcache` BPF maps, respectively.
+
+### RPC Methods
+
+Method : UpdateMappings
+
+_rpc UpdatesMappings(steam FQDNMapping) returns (Result){}_
+Request :
+```
+message FQDNMapping {
+    string FQDN = 1;
+    repeated bytes IPS = 2;
+    int32 TTL = 3;
+    bytes client_ip = 4;
+    int32 response_code = 5;
+}
+```
+Response :
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : UpdatesDNSRules
+
+_rpc UpdatesDNSRules(stream DNSPolicyRules) returns (Result){}_
+Request :
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  string selector_string = 1;
+  repeated FQDNSelector port_rules= 2;
+  repeated string match_labels = 3;
+  repeated uint32 selections = 4;
+}
+
+message DNSPolicyRules {
+  uint64 endpoint_id = 1;
+  uint32 port = 2;
+  repeated DNSPolicyRule rules = 3;
+}
+```
+
+Response :
+```
+message Result  {
+    message string = 1;
+}
+```
+
+### Tracking policy updates to SDP instances
+
+Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
+Since policy revision numbers are reset when agent restarts, we need to unconditionally send policy updates to SDP on agent restart.
+
+## Impacts / Key Questions
+
+### Getting the DNS Rules
+
+We need the DNS Rules for the Standalone DNS proxy to enforce the L7 DNS policy. The policy are the source of turth for the rules and are propagated to the agent when we apply the policy. We explore the options to get the DNS rules for the DNS proxy.
+
+#### Running the GRPC sever in the agent
+
+We can run a gRPC server in the agent to serve the DNS rules to the DNS proxy. SDP will be responsible for creating the connection with the agent. And once SDP establish a connection agent can keep track of the stream and send the DNS rules to the SDP. Agent can then reuse the same stream to send updates in the DNS rules.
+In case, cilium agent is still not up, SDP will keep trying to connect to the agent until the connection is established.
+
+##### Pros
+
+* SDP instances has the responsibility to connect to the agent.
```

Comment:
We thought about different ways of sending the DNS rules to SDP.
One way was sending rules to a particular `ip:port` and in that case, Cilium agent has the responsibility to find if SDP is running and then send out the rules.

In this case, where SDP is connecting to the agent. It is the responsibility of the SDP to make sure it connects to agent rather than other way around.
This reduces the overhead for the cilium agent to look out for processes(in case of multiple instances of proxies running) and send them rules.

Reply:
I read "Simpler to configure and scale out SDP instances since the agent only needs a socket to be configured and does not need to explicitly reference the SDP instances". That's a pro :+1: